### DISCUSSION TITLE
How to start seaweedfs s3 with one bucket?

### DISCUSSION THREAD

**Discussion by Jeeppler** [seaweedfs/seaweedfs#2461] [General]
I am new to seaweedfs. My objective is to use seaweedfs to for a single S3 bucket. This is all I need. I am running seaweedfs in my own container. I need to upload very small files (approx. 5 MB) and only a few of them.

Currently, I am trying to start seaweedfs with the following command:

~~~
weed server -dir=/storage -master.volumeSizeLimitMB=1024 -master.volumePreallocate=false -s3 -s3.port 9000 -s3.config "/home/$USER/s3_config.json"
~~~

The S3 config looks like this:

~~~
/home/seaweedfs/s3_config.json 
{
    "identities": [
        {
            "name": "me",
            "credentials": [
                {
                "accessKey": "any",
                "secretKey": "anylongkey"
                }
            ],
            "actions": [
                "Read",
                "Write",
                "List",
                "Tagging",
                "Admin"
            ]
        }
    ]
}
~~~

1. this command does not create volumes right away, it takes a long time for the volumes to be created.
My disk stats are: 

    ~~~
    Path 	Disk 	Total 	Free 	Usage
     /storage 		196.36 GiB 	20.65 GiB 	89.48%
    ~~~
2. SeaweedFS still shows in the logs:

    ~~~
    I1119 20:59:32    17 volume.go:360] Start Seaweed volume server 30GB 2.78 100c654ec3b00783ec1cdf8fa71f31a63d584210 at :8080
    ~~~

3. The `/buckets/<bucket-name>` folder is not getting created.
4. I am unable to connect to S3 using `s3cmd`.

Currently, I am pretty lost and appreciate any help.

**Comment by chrislusf** [comment]
Could you please try to add “=“ in the command line?

**Comment by Jeeppler** [comment]
I tried this: weed server -dir=/storage -master.volumeSizeLimitMB=1024 -master.volumePreal
locate=false -s3 -s3.port=9000 -s3.config="/home/$USER/s3_config.json". The problems remain.

My output to standard out:

~~~
weed server -dir=/storage -master.volumeSizeLimitMB=1024 -master.volumePreal
locate=false -s3 -s3.port=9000 -s3.config="/home/$USER/s3_config.json"
I1122 15:46:48    33 master.go:170] current: 10.42.43.7:9333 peers:
I1122 15:46:48    33 file_util.go:23] Folder /storage Permission: -rwxr-xr-x
I1122 15:46:48    33 master.go:170] current: 10.42.43.7:9333 peers:10.42.43.7:9333
I1122 15:46:48    33 master_server.go:119] Volume Size Limit is 1024 MB
I1122 15:46:48    33 master_server.go:213] adminScripts:
I1122 15:46:48    33 master.go:125] Start Seaweed Master 30GB 2.78 100c654ec3b00783ec1cdf8fa71f31a63d584210 at :9333
I1122 15:46:48    33 raft_server.go:71] Starting RaftServer with 10.42.43.7:9333
I1122 15:46:48    33 file_util.go:23] Folder /storage Permission: -rwxr-xr-x
I1122 15:46:48    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:46:48    33 raft_server.go:131] current cluster leader: 
I1122 15:46:50    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:46:50    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:51    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:52    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:46:52    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:53    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:54    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:46:54    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:55    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:56    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:46:56    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:57    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:46:58    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:59    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:46:59    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:47:00    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:01    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:01    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:47:02    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:03    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:03    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:47:04    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:05    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:05    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:47:06    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:06    33 volume_grpc_client_to_master.go:41] checkWithMaster 10.42.43.7:9333: get master 10.42.43.7:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:19333: connect: connection refused"
I1122 15:47:06    33 master.go:148] Start Seaweed Master 30GB 2.78 100c654ec3b00783ec1cdf8fa71f31a63d584210 grpc server at :19333
I1122 15:47:07    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:08    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:08    33 masterclient.go:79] No existing leader found!
I1122 15:47:08    33 raft_server.go:158] Initializing new cluster
I1122 15:47:08    33 master_server.go:162] leader change event:  => 10.42.43.7:9333
I1122 15:47:08    33 master_server.go:164] [ 10.42.43.7:9333 ] 10.42.43.7:9333 becomes leader.
I1122 15:47:08    33 disk_location.go:180] Store started on dir: /storage with 0 volumes max 8
I1122 15:47:08    33 disk_location.go:183] Store started on dir: /storage with 0 ec shards
I1122 15:47:08    33 volume.go:360] Start Seaweed volume server 30GB 2.78 100c654ec3b00783ec1cdf8fa71f31a63d584210 at :8080
I1122 15:47:08    33 volume_grpc_client_to_master.go:50] Volume server start with seed master nodes: [10.42.43.7:9333]
I1122 15:47:08    33 volume_grpc_client_to_master.go:107] Heartbeat to: 10.42.43.7:9333
I1122 15:47:08    33 node.go:222] topo adds child DefaultDataCenter
I1122 15:47:08    33 node.go:222] topo:DefaultDataCenter adds child DefaultRack
I1122 15:47:08    33 node.go:222] topo:DefaultDataCenter:DefaultRack adds child 10.42.43.7:8080
I1122 15:47:08    33 node.go:222] topo:DefaultDataCenter:DefaultRack:10.42.43.7:8080 adds child 
I1122 15:47:08    33 master_grpc_server.go:70] added volume server 0: 10.42.43.7:8080
I1122 15:47:09    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:10    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:11    33 filer_server.go:120] default to create filer store dir in /storage/filerldb2
I1122 15:47:11    33 master_grpc_server.go:262] + client filer@10.42.43.7:8888
I1122 15:47:11    33 leveldb2_store.go:42] filer store leveldb2 dir: /storage/filerldb2
I1122 15:47:11    33 file_util.go:23] Folder /storage/filerldb2 Permission: -rwxr-xr-x
I1122 15:47:11    33 s3.go:162] wait to connect to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:11    33 filer.go:121] create filer.store.id to -1291683835
I1122 15:47:11    33 configuration.go:28] configured filer store to leveldb2
I1122 15:47:11    33 filer.go:89] the cluster has 1 filers
I1122 15:47:11    33 filer.go:222] Start Seaweed Filer 30GB 2.78 100c654ec3b00783ec1cdf8fa71f31a63d584210 at 10.42.43.7:8888
I1122 15:47:11    33 meta_aggregator.go:100] connecting to peer filer 10.42.43.7:8888: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.43.7:18888: connect: connection refused"
I1122 15:47:11    33 master_grpc_server.go:262] + client master@10.42.43.7:9333
I1122 15:47:12    33 s3.go:158] S3 read filer buckets dir: /buckets
I1122 15:47:12    33 s3.go:165] connected to filer 10.42.43.7:8888 grpc address 10.42.43.7:18888
I1122 15:47:12    33 s3.go:201] Start Seaweed S3 API Server 30GB 2.78 100c654ec3b00783ec1cdf8fa71f31a63d584210 at http port 9000
I1122 15:47:12    33 filer_grpc_server_sub_meta.go:242] + listener s3@10.42.43.7:58956
I1122 15:47:12    33 filer_grpc_server_sub_meta.go:31]  s3@10.42.43.7:58956 starts to subscribe /etc/iam/identity.json from 2021-11-22 15:47:12.058443497 +0000 UTC
I1122 15:47:12    33 filer_grpc_server_sub_meta.go:242] + listener filer:10.42.43.7:8888@10.42.43.7:58956
I1122 15:47:12    33 filer_grpc_server_sub_meta.go:89]  filer:10.42.43.7:8888@10.42.43.7:58956 local subscribe / from 2021-11-22 15:46:11.089785608 +0000 UTC
I1122 15:47:12    33 filer_grpc_server_sub_meta.go:101] read on disk filer:10.42.43.7:8888@10.42.43.7:58956 local subscribe / from 2021-11-22 15:46:11.089785608 +0000 UTC
I1122 15:47:12    33 filer_grpc_server_sub_meta.go:117] read in memory filer:10.42.43.7:8888@10.42.43.7:58956 local subscribe / from 2021-11-22 15:46:11.089785608 +0000 UTC
~~~

**Comment by chrislusf** [comment]
The buckets are lazily created, when the data is actually written.

I do not know whether the `"/home/$USER/s3_config.json"` will be received correctly. Can you try to spell it out?

**Comment by Jeeppler** [comment]
`/home/$USER/s3_config.json` looks like this:

~~~
{
    "identities": [
        {
            "name": "me",
            "credentials": [
                {
                "accessKey": "any",
                "secretKey": "anylongkey"
                }
            ],
            "actions": [
                "Read",
                "Write",
                "List",
                "Tagging",
                "Admin"
            ]
        }
    ]
}
~~~

Basically, we want to replace [MinIO with SeaweedFS](Daimler/sechub/issues/895). You can have a look at all the files here: https://github.com/Daimler/sechub/pull/901/.

**Reply by chrislusf** [reply]
sorry, I mean spelling out the `$USER` part so that the CLI can resolve to the right file.

**Reply by Jeeppler** [reply]
the `$USER` variable is set in the Dockerfile to `seaweedfs`. If I put the value into the command like this:  `/home/seaweedfs/s3_config.json`.

**Reply by chrislusf** [reply]
can you try to use `/home/seaweedfs/s3_config.json` directly?

**Reply by Jeeppler** [reply]
Yes, I did. However, that was not the problem. I don't know exactly how, but my configuration works now.

Currently, I struggle with `s3cmd`. I am unable to connect to the seaweedfs server (S3) using `s3cmd`.

**Reply by chrislusf** [reply]
s3cmd with SeaweedFS https://github.com/chrislusf/seaweedfs/wiki/s3cmd-with-SeaweedFS

**Reply by Jeeppler** [reply]
Thanks for your help.

**Comment by Jeeppler** [comment]
@chrislusf is there a way to create buckets non lazy. Meaning initialize seaweedfs with buckets?

**Reply by chrislusf** [reply]
you can create `/buckets/<bucket_name>` folder directly.

or use this
```
echo "s3.bucket.create -name xxx" | weed shell
```

**Reply by Jeeppler** [reply]
Creating the `/buckets/<bucket_name>` did not work. 

This works:

~~~
echo "s3.bucket.create -name xxx" | weed shell
~~~

But I have to wait until the server started (approx. 25 seconds).

**Reply by chrislusf** [reply]
why lazy bucket creation does not work for you?

**Reply by Jeeppler** [reply]
My problem is, the application relying on the bucket throws an exception, if the bucket is not created upfront.
